File: apps/ios/NOVA/Models/GitHubIntegration.swift
import Foundation

/// The owner's GitHub webhook, as the settings screen sees it.
///
/// There is no secret here. It exists in exactly one response -- the one
/// that creates or rotates the integration -- and never again, so a screen
/// that reloads cannot show it, and neither can a screenshot taken later.
struct GitHubIntegration: Codable, Identifiable, Equatable, Sendable {
    let id: UUID
    /// "owner/name", or nil to react to any repository the hook covers.
    let repository: String?
    let enabled: Bool
    /// Absolute when the server knows its public name; otherwise a path,
    /// and the screen says so.
    let webhookUrl: String
    let lastDeliveryAt: Date?
    /// e.g. "workflow_run:success". The answer to "is this wired up?".
    let lastEvent: String?
    let createdAt: Date

    var isAbsoluteURL: Bool { webhookUrl.hasPrefix("http") }
}

/// The creation response: the integration plus its secret, once.
struct GitHubIntegrationCreated: Codable, Equatable, Sendable {
    let id: UUID
    let repository: String?
    let enabled: Bool
    let webhookUrl: String
    let lastDeliveryAt: Date?
    let lastEvent: String?
    let createdAt: Date
    let secret: String
    let contentType: String

    var integration: GitHubIntegration {
        GitHubIntegration(
            id: id, repository: repository, enabled: enabled, webhookUrl: webhookUrl,
            lastDeliveryAt: lastDeliveryAt, lastEvent: lastEvent, createdAt: createdAt
        )
    }
}

struct CreateGitHubIntegrationRequest: Encodable, Sendable {
    var repository: String?
}

struct UpdateGitHubIntegrationRequest: Encodable, Sendable {
    var repository: String?
    var enabled: Bool?
}

File: services/api/src/nova/models/github_integration.py
"""A user's GitHub webhook integration.

One per user. The secret is what GitHub signs deliveries with, so it has to
be stored as it is -- HMAC verification needs the shared key, not a hash of
it. It is generated here, shown to the user exactly once at creation, and
never returned again; rotating it is the only way to see a new one. That is
the same trade the device token in the firmware's NVS makes, and the same
mitigation applies: it is revocable in one call.
"""

from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import Boolean, DateTime, ForeignKey, String
from sqlalchemy.orm import Mapped, mapped_column

from nova.db.base import Base, TimestampMixin, UUIDPrimaryKeyMixin


class GitHubIntegration(Base, UUIDPrimaryKeyMixin, TimestampMixin):
    __tablename__ = "github_integrations"

    user_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), nullable=False, unique=True, index=True
    )
    # Shared with GitHub; verifies every delivery. See the module docstring.
    webhook_secret: Mapped[str] = mapped_column(String(128), nullable=False)
    # "owner/name". When set, deliveries from any other repository are
    # acknowledged and ignored -- a webhook is per-repository on GitHub's
    # side, but an organisation-level hook can carry many.
    repository: Mapped[str | None] = mapped_column(String(200), nullable=True)
    enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

    last_delivery_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    # e.g. "workflow_run:success". For the settings screen, so "is this
    # wired up?" has an answer without reading server logs.
    last_event: Mapped[str | None] = mapped_column(String(64), nullable=True)
